formats/cbor/commonMain/src/kotlinx/serialization/cbor/internal/CborElementSerializers.kt

```diff
@@ -0,0 +1,304 @@
+        MapSerializer(CborElementSerializer, CborElementSerializer).serialize(encoder, value)
+    }
+
+    override fun deserialize(decoder: Decoder): CborMap {
+        decoder.asCborDecoder()
+        return CborMap(MapSerializer(CborElementSerializer, CborElementSerializer).deserialize(decoder))
+    }
+}
+
+/**
+ * Serializer object providing [SerializationStrategy] and [DeserializationStrategy] for [CborList].
+ * It can only be used by with [Cbor] format an its input ([CborDecoder] and [CborEncoder]).
+ */
+public object CborListSerializer : KSerializer<CborList> {
+    private object CborListDescriptor : SerialDescriptor by ListSerializer(CborElementSerializer).descriptor {
+        @ExperimentalSerializationApi
+        override val serialName: String = "kotlinx.serialization.cbor.CborList"
+    }
+
+    override val descriptor: SerialDescriptor = CborListDescriptor
+
+    override fun serialize(encoder: Encoder, value: CborList) {
+        val cborEncoder = encoder.asCborEncoder()
+        cborEncoder.encodeTags(value)
+        ListSerializer(CborElementSerializer).serialize(encoder, value)
+    }
+
+    override fun deserialize(decoder: Decoder): CborList {
+        decoder.asCborDecoder()
+        return CborList(ListSerializer(CborElementSerializer).deserialize(decoder))
+    }
+}
+
+
+internal fun Decoder.asCborDecoder(): CborDecoder = this as? CborDecoder
+    ?: throw IllegalStateException(
+        "This serializer can be used only with Cbor format." +
+            "Expected Decoder to be CborDecoder, got ${this::class}"
+    )
+
+/*need to expose writer to access encodeTag()*/
+internal fun Encoder.asCborEncoder() = this as? CborWriter
+    ?: throw IllegalStateException(
+        "This serializer can be used only with Cbor format." +
+            "Expected Encoder to be CborEncoder, got ${this::class}"
+    )
+
+/**
+ * Returns serial descriptor that delegates all the calls to descriptor returned by [deferred] block.
+ * Used to resolve cyclic dependencies between recursive serializable structures.
+ */
+@OptIn(ExperimentalSerializationApi::class)
+private fun defer(deferred: () -> SerialDescriptor): SerialDescriptor = object : SerialDescriptor {
+    private val original: SerialDescriptor by lazy(deferred)
+
+    override val serialName: String
+        get() = original.serialName
+    override val kind: SerialKind
+        get() = original.kind
+    override val elementsCount: Int
+        get() = original.elementsCount
+
+    override fun getElementName(index: Int): String = original.getElementName(index)
+    override fun getElementIndex(name: String): Int = original.getElementIndex(name)
+    override fun getElementAnnotations(index: Int): List<Annotation> = original.getElementAnnotations(index)
+    override fun getElementDescriptor(index: Int): SerialDescriptor = original.getElementDescriptor(index)
+    override fun isElementOptional(index: Int): Boolean = original.isElementOptional(index)
+}
```

Comment:
Rather than this, it may be better to have the descriptor property itself be lazy, rather than each of its elements.